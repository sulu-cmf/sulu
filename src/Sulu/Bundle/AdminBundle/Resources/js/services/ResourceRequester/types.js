// @flow
export type ListOptions = {
    page: number,
    limit: ?number,
    locale?: ?string,
};
