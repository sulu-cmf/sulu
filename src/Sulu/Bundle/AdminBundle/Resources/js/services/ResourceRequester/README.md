The `ResourceRequester` is built for talking to APIs as they are built in Sulu. It already knows how this API is built
and sends the correct requests resp. handles the responses correctly.

There are a few methods that can be used to talk to these APIs. They all share the same methods, and the first
parameter is the `resourceKey`, which defines the entity.

```javascript static
ResourceRequester.getList('contacts', {page: 2})
    .then((response) => {
        // outputs the second page of contacts
        console.log(response._embedded.contacts);
    });

// delete the snippet with the ID 6
ResourceRequester.delete('snippets', 6);
```
