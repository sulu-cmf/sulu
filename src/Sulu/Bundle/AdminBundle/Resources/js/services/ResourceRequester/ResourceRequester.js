// @flow
import Requester from '../Requester';
import resourceMetadataStore from '../../stores/ResourceMetadataStore';
import type {ListOptions} from './types';

export default class ResourceRequester {
    static buildQueryString(queryOptions: ?Object) {
        const options = queryOptions;
        if (!options) {
            return '';
        }

        const searchParameters = new URLSearchParams();
        Object.keys(options).forEach((key) => {
            if (options[key] === undefined) {
                return;
            }

            searchParameters.set(key, options[key]);
        });

        return '?' + searchParameters.toString().replace(/%2C/gi, ',');
    }

    static get(resourceKey: string, id: ?number | string, queryOptions: ?Object) {
        const endpoint = resourceMetadataStore.getEndpoint(resourceKey);
        return Requester.get(endpoint + (id ? '/' + id : '') + ResourceRequester.buildQueryString(queryOptions));
    }

    static post(resourceKey: string, data: Object, queryOptions: ?Object) {
        const endpoint = resourceMetadataStore.getEndpoint(resourceKey);
        return Requester.post(endpoint + ResourceRequester.buildQueryString(queryOptions), data);
    }

    static postWithId(resourceKey: string, id: number | string, data: Object, queryOptions: ?Object) {
        const endpoint = resourceMetadataStore.getEndpoint(resourceKey);
        return Requester.post(endpoint + '/' + id + ResourceRequester.buildQueryString(queryOptions), data);
    }

    static put(resourceKey: string, id: number | string, data: Object, queryOptions: ?Object) {
        const endpoint = resourceMetadataStore.getEndpoint(resourceKey);
        return Requester.put(endpoint + '/' + id + ResourceRequester.buildQueryString(queryOptions), data);
    }

    static getList(resourceKey: string, options: ListOptions) {
        const endpoint = resourceMetadataStore.getEndpoint(resourceKey);
        const queryOptions = {...options, flat: true};

        return Requester.get(endpoint + ResourceRequester.buildQueryString(queryOptions));
    }

    static delete(resourceKey: string, id: number | string, queryOptions: ?Object) {
        const endpoint = resourceMetadataStore.getEndpoint(resourceKey);
        return Requester.delete(endpoint + '/' + id + ResourceRequester.buildQueryString(queryOptions));
    }
}
