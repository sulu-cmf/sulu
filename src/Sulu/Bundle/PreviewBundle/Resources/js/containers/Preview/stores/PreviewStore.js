// @flow
import queryString from 'query-string';
import {action, computed, observable} from 'mobx';
import Requester from 'sulu-admin-bundle/services/Requester';
import type {PreviewRouteName} from './../types';

const generateRoute = (name: PreviewRouteName, options: Object): string => {
    const query = queryString.stringify(options);
    if (query.length === 0) {
        return PreviewStore.routes[name];
    }

    return PreviewStore.routes[name] + '?' + query;
};

export default class PreviewStore {
    static routes: {[PreviewRouteName]: string} = {};

    resourceKey: string;
    id: ?string | number;
    locale: string;
    webspace: string;

    @observable token: ?string;

    constructor(resourceKey: string, id: ?string | number, locale: string, webspace: string) {
        this.resourceKey = resourceKey;
        this.id = id;
        this.locale = locale;
        this.webspace = webspace;
    }

    @computed get starting() {
        return !this.token;
    }

    @computed get renderRoute() {
        return generateRoute('render', {
            webspace: this.webspace,
            locale: this.locale,
            token: this.token,
        });
    }

    @action setToken = (token: ?string) => {
        this.token = token;
    };

    start(): Promise<*> {
        const route = generateRoute('start', {
            provider: this.resourceKey,
            locale: this.locale,
            id: this.id,
        });

        return Requester.get(route).then((response) => {
            this.setToken(response.token);
        });
    }

    update(data: Object): Promise<string> {
        const route = generateRoute('update', {
            locale: this.locale,
            webspace: this.webspace,
            token: this.token,
        });

        return Requester.post(route, {data: data}).then((response) => {
            return response.content;
        });
    }

    updateContext(type: string): Promise<string> {
        const route = generateRoute('update-context', {
            webspace: this.webspace,
            token: this.token,
        });

        return Requester.post(route, {context: {template: type}}).then((response) => {
            return response.content;
        });
    }

    stop(): Promise<*> {
        const route = generateRoute('stop', {token: this.token});

        return Requester.get(route).then(this.setToken(null));
    }
}
