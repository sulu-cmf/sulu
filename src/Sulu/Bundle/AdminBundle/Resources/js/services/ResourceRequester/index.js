// @flow
import ResourceRequester from './ResourceRequester';

export default ResourceRequester;
