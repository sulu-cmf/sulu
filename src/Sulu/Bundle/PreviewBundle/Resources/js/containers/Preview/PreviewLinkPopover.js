// @flow
import React from 'react';
import {action, computed, observable} from 'mobx';
import {observer} from 'mobx-react';
import copyToClipboard from 'copy-to-clipboard';
import Button from 'sulu-admin-bundle/components/Button';
import {ResourceRequester} from 'sulu-admin-bundle/services';
import {transformDateForUrl} from 'sulu-admin-bundle/utils/Date';
import {translate} from 'sulu-admin-bundle/utils';
import PreviewStore from './stores/PreviewStore';
import previewLinkStyles from './preview-link.scss';
import type {PreviewLink} from './types';

type Props = {|
    previewStore: PreviewStore,
|};

@observer
class PreviewLinkPopover extends React.Component<Props> {
    @observable previewLink: ?PreviewLink;
    @observable loading: boolean = false;
    @observable generating: boolean = false;
    @observable copying: boolean = false;

    componentDidMount() {
        this.loadPreviewLink();
    }

    @action loadPreviewLink() {
        const {
            previewStore,
        } = this.props;

        this.loading = true;
        ResourceRequester.get('preview_links', {
            resourceKey: previewStore.resourceKey,
            resourceId: previewStore.id,
            locale: previewStore.locale,
        }).then(action((previewLink) => {
            this.previewLink = previewLink;
            this.loading = false;
        })).catch(action((error) => {
            if (error.status !== 404) {
                return Promise.reject(error);
            }

            this.loading = false;
        }));
    }

    @action handleGenerateClick = () => {
        const {
            previewStore,
        } = this.props;

        this.generating = true;
        ResourceRequester.post('preview_links', {}, {
            action: 'generate',
            resourceKey: previewStore.resourceKey,
            resourceId: previewStore.id,
            locale: previewStore.locale,
            webspaceKey: previewStore.webspace,
            segmentKey: previewStore.segment,
            targetGroupId: previewStore.targetGroup,
            dateTime: previewStore.dateTime && transformDateForUrl(previewStore.dateTime),
        }).then(action((previewLink) => {
            this.previewLink = previewLink;
        })).finally(action(() => this.generating = false));
    };

    handleRevokeClick = (event: SyntheticEvent<HTMLFormElement>) => {
        event.preventDefault();

        const {
            previewStore,
        } = this.props;

        ResourceRequester.post('preview_links', {}, {
            action: 'revoke',
            resourceKey: previewStore.resourceKey,
            resourceId: previewStore.id,
            locale: previewStore.locale,
        }).then(action(() => {
            this.previewLink = null;
        }));
    };

    @action handleCopyClick = () => {
        this.copying = true;
        setTimeout(action(() => this.copying = false), 125);

        copyToClipboard(this.link);
    };

    @computed get link() {
        if (!this.previewLink) {
            return '';
        }

        return PreviewStore.endpoints['preview-link'].replace(':token', this.previewLink.token);
    }

    render() {
        if (this.loading) {
            return null;
        }

        return (
            <div className={previewLinkStyles.container}>
                {this.previewLink && (
                    <React.Fragment>
                        <div>
                            <label className={previewLinkStyles.label}>
                                {translate('sulu_preview.copy_preview_link')}
                            </label>
                            <div className={previewLinkStyles.inputContainer}>
                                <input
                                    className={previewLinkStyles.input}
                                    readOnly={true}
                                    value={this.link}
                                />

                                <Button
                                    className={previewLinkStyles.copyButton}
                                    loading={this.copying}
                                    onClick={this.handleCopyClick}
                                    skin="primary"
                                >
                                    {translate('sulu_preview.copy')}
                                </Button>
                            </div>
                        </div>
                        <div className={previewLinkStyles.revoke}>
                            <button
                                className={previewLinkStyles.revokeButton}
                                onClick={this.handleRevokeClick}
                                type="button"
                            >
                                {translate('sulu_preview.revoke')}
                            </button>
                        </div>
                    </React.Fragment>
                )}
                {!this.previewLink && (
                    <React.Fragment>
                        <Button
                            loading={this.generating}
                            onClick={this.handleGenerateClick}
                            skin="primary"
                        >
                            {translate('sulu_preview.generate_link')}
                        </Button>
                    </React.Fragment>
                )}
            </div>
        );
    }
}

export default PreviewLinkPopover;
