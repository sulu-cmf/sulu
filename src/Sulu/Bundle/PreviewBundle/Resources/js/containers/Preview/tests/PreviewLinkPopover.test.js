// @flow
import React from 'react';
import {observable} from 'mobx';
import {shallow} from 'enzyme';
import copyToClipboard from 'copy-to-clipboard';
import ResourceRequester from 'sulu-admin-bundle/services/ResourceRequester';
import PreviewStore from '../stores/PreviewStore';
import PreviewLinkPopover from '../PreviewLinkPopover';

jest.mock('copy-to-clipboard', () => jest.fn());

jest.mock('sulu-admin-bundle/services/ResourceRequester', () => ({
    get: jest.fn(),
    post: jest.fn(),
}));

jest.mock('../stores/PreviewStore', () => jest.fn(function(resourceKey, id, locale) {
    this.resourceKey = resourceKey;
    this.id = id;
    this.locale = locale.get();
}));

jest.mock('sulu-admin-bundle/utils', () => ({
    translate: jest.fn((key) => key),
}));

beforeEach(() => {
    jest.resetModules();

    PreviewStore.endpoints = {
        'preview-link': '/admin/p/:token',
    };
});

test('Render popover when preview link is available and copy link to clipboard', () => {
    const promise = Promise.resolve({
        token: '123-123-123',
    });
    ResourceRequester.get.mockReturnValue(promise);

    const previewStore = new PreviewStore(
        'pages',
        '123-123-123',
        observable.box('de'),
        'sulu_io',
        undefined
    );
    const component = shallow(<PreviewLinkPopover previewStore={previewStore} />);

    return promise.then(() => {
        expect(component).toMatchSnapshot();

        expect(ResourceRequester.get).toBeCalledWith('preview_links', {
            resourceKey: 'pages',
            resourceId: '123-123-123',
            locale: 'de',
        });

        component.instance().handleCopyClick();

        expect(copyToClipboard).toBeCalledWith('/admin/p/123-123-123');
    });
});

test('Render popover when no link is available', (done) => {
    const promise = Promise.reject({
        status: 404,
    });
    ResourceRequester.get.mockReturnValue(promise);

    const previewStore = new PreviewStore(
        'pages',
        '123-123-123',
        observable.box('de'),
        'sulu_io',
        undefined
    );
    const component = shallow(<PreviewLinkPopover previewStore={previewStore} />);

    setTimeout(() => {
        expect(component).toMatchSnapshot();

        expect(ResourceRequester.get).toBeCalledWith('preview_links', {
            resourceKey: 'pages',
            resourceId: '123-123-123',
            locale: 'de',
        });

        done();
    });
});

test('Generate link', (done) => {
    const promise = Promise.reject({
        status: 404,
    });
    ResourceRequester.get.mockReturnValue(promise);

    const previewStore = new PreviewStore(
        'pages',
        '123-123-123',
        observable.box('de'),
        'sulu_io',
        undefined
    );
    const component = shallow(<PreviewLinkPopover previewStore={previewStore} />);

    setTimeout(() => {
        const promise = Promise.resolve({
            token: '123-123-123',
        });
        ResourceRequester.post.mockReturnValue(promise);

        component.instance().handleGenerateClick();

        setTimeout(() => {
            expect(component).toMatchSnapshot();

            expect(ResourceRequester.post).toBeCalledWith('preview_links', {}, {
                action: 'generate',
                dateTime: undefined,
                resourceKey: 'pages',
                resourceId: '123-123-123',
                locale: 'de',
                segmentKey: undefined,
                targetGroupId: undefined,
                webspaceKey: undefined,
            });

            done();
        });
    });
});

test('Revoke Link', (done) => {
    const promise = Promise.resolve({
        token: '123-123-123',
    });
    ResourceRequester.get.mockReturnValue(promise);

    const previewStore = new PreviewStore(
        'pages',
        '123-123-123',
        observable.box('de'),
        'sulu_io',
        undefined
    );
    const component = shallow(<PreviewLinkPopover previewStore={previewStore} />);

    promise.then(() => {
        const promise = Promise.resolve();
        ResourceRequester.post.mockReturnValue(promise);

        component.instance().handleRevokeClick({preventDefault: jest.fn()});

        expect(ResourceRequester.post.mock.calls[0][0]).toBe('preview_links');
        expect(ResourceRequester.post.mock.calls[0][1]).toStrictEqual({});
        expect(ResourceRequester.post.mock.calls[0][2]).toStrictEqual({
            action: 'revoke',
            resourceKey: 'pages',
            resourceId: '123-123-123',
            locale: 'de',
        });

        done();
    });
});
