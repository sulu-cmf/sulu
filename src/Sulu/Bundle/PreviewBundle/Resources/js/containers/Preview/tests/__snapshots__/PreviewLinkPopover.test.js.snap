// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate link 1`] = `
<div
  className="container"
>
  <div>
    <label
      className="label"
    >
      sulu_preview.copy_preview_link
    </label>
    <div
      className="inputContainer"
    >
      <input
        className="input"
        readOnly={true}
        value="/admin/p/123-123-123"
      />
      <Button
        active={false}
        className="copyButton"
        disabled={false}
        loading={false}
        onClick={[Function]}
        showDropdownIcon={false}
        size="large"
        skin="primary"
        type="button"
      >
        sulu_preview.copy
      </Button>
    </div>
  </div>
  <div
    className="revoke"
  >
    <button
      className="revokeButton"
      onClick={[Function]}
      type="button"
    >
      sulu_preview.revoke
    </button>
  </div>
</div>
`;

exports[`Render popover when no link is available 1`] = `
<div
  className="container"
>
  <Button
    active={false}
    disabled={false}
    loading={false}
    onClick={[Function]}
    showDropdownIcon={false}
    size="large"
    skin="primary"
    type="button"
  >
    sulu_preview.generate_link
  </Button>
</div>
`;

exports[`Render popover when preview link is available and copy link to clipboard 1`] = `
<div
  className="container"
>
  <div>
    <label
      className="label"
    >
      sulu_preview.copy_preview_link
    </label>
    <div
      className="inputContainer"
    >
      <input
        className="input"
        readOnly={true}
        value="/admin/p/123-123-123"
      />
      <Button
        active={false}
        className="copyButton"
        disabled={false}
        loading={false}
        onClick={[Function]}
        showDropdownIcon={false}
        size="large"
        skin="primary"
        type="button"
      >
        sulu_preview.copy
      </Button>
    </div>
  </div>
  <div
    className="revoke"
  >
    <button
      className="revokeButton"
      onClick={[Function]}
      type="button"
    >
      sulu_preview.revoke
    </button>
  </div>
</div>
`;
